### PR TITLE
add click outside hook

### PR DESCRIPTION
**Pull request by infothien**

### CODE REVIEW

File: example/index.res [hunk 1]
```diff
@@ -1,3 +1,39 @@
...
+  let make = () => {
+    let inputRef = React.useRef(Js.Nullable.null)
+
+    let listRef = React.useRef(Js.Nullable.null)
+
+    let (isClickedOutside, setIsClickedOutside) = React.useState(_ => true)
+
+    let (text, setText) = React.useState(_ => "")
+
+    ClickOutsideHook.use(~refs=[inputRef, listRef], () => setIsClickedOutside(_ => true))
```
Comment (gaku-sei) [inline example/index.res]: I wonder, for the sake of keeping the exemple as straightforward as possible, shouldn't we have only 1 ref

File: src/ClickOutsideHook.res [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+If it is yes do nothing
+Otherwise call function onClickOutside
+")
+let use = (~refs, ~onClick=true, ~onTouch=true, onClickOutside) => {
+  let listener = element =>
+    if (
+      !(
+        refs
+        ->Belt_Array.keepMap(({React.current: current}) => current->Js.Nullable.toOption)
+        ->Belt_Array.some(Webapi.Dom.Element.contains(element))
```
Comment (gaku-sei) [inline src/ClickOutsideHook.res]: Let's use only one loop here and keep `some` 👍
We can also use `Js.Nullable.map` directly in that case

File: example/index.res [hunk 3]
```diff
@@ -1,3 +1,39 @@
+module ClickOutside = {
+  @react.component
+  let make = () => {
+    let inputRef = React.useRef(Js.Nullable.null)
+
+    let listRef = React.useRef(Js.Nullable.null)
+
+    let (isClickedOutside, setIsClickedOutside) = React.useState(_ => true)
```
Comment (gaku-sei) [inline example/index.res]: The wording is a bit confusing here, could it be `dropdownIsOpen` or `fruitsDropdownIsOpen`?

File: src/ClickOutsideHook.res [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  ")
+  let touchStartListener = event =>
+    event
+    ->Webapi.Dom.TouchEvent.target
+    ->Webapi.Dom.EventTarget.unsafeAsElement
+    ->{onTouch ? listener : ignore}
+
+  ReHooks.WindowListenersHook.MouseDown.use0(mouseDownListener)
+
+  ReHooks.WindowListenersHook.TouchStart.use0(touchStartListener)
```
Comment (gaku-sei) [inline src/ClickOutsideHook.res]: Let's remove the ReHooks module 👍

File: src/ClickOutsideHook.res [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    event
+    ->Webapi.Dom.MouseEvent.target
+    ->Webapi.Dom.EventTarget.unsafeAsElement
+    ->{onClick ? listener : ignore}
+
+  @ocaml.doc("
+  Takes a touch event and return its target as an Html element
+  and apply listener when touch start
+  ")
+  let touchStartListener = event =>
```
Comment (gaku-sei) [inline src/ClickOutsideHook.res:30]: Since these functions are used once and only in this context maybe we can use a simple comment instead of ocaml.doc?

**Comment by infothien** [comment]
Thank you @gaku-sei -san for your comments, I fixed all and force pushed it

File: src/WindowListenersHook.res [hunk 6]
```diff
@@ -87,6 +89,8 @@ let make = (
         Some(() => Webapi.Dom.window->removeEventListener(listener, _))
       }
 
+      let use = listener => React.useEffect(() => makeListener(listener))
```
Comment (gaku-sei) [inline src/WindowListenersHook.res:92]: I wonder why this one was absent 🤔

**Comment by infothien** [comment]
Thank you for the changes, LGTM

**Comment by gaku-sei** [comment]
Let's goooooo

**Pull request merged**